diff --git a/home/plasterr/Git/test/kernel_pixel/aosp/include/linux/usb/tcpm.h b/home/plasterr/Git/graphene_pixel/kernel_pixel/aosp/include/linux/usb/tcpm.h
index ab7ca87..9d83aed 100644
--- a/aosp/include/linux/usb/tcpm.h
+++ b/aosp/include/linux/usb/tcpm.h
@@ -161,6 +161,10 @@ struct tcpm_port;
 struct tcpm_port *tcpm_register_port(struct device *dev, struct tcpc_dev *tcpc);
 void tcpm_unregister_port(struct tcpm_port *port);
 
+int tcpm_update_sink_capabilities(struct tcpm_port *port, const u32 *pdo,
+				  unsigned int nr_pdo,
+				  unsigned int operating_snk_mw);
+
 void tcpm_vbus_change(struct tcpm_port *port);
 void tcpm_cc_change(struct tcpm_port *port);
 void tcpm_sink_frs(struct tcpm_port *port);
